src/autoscaler/collection/TSDCache_test.go

```diff
@@ -0,0 +1,199 @@
+			})
+		})
+		Context("when data put to cache execeed the capacity", func() {
+			BeforeEach(func() {
+				capacity = 3
+			})
+			It("caches latest data in ascending order", func() {
+				cache.Put(TestTSD{20, nil})
+				Expect(cache.String()).To(Equal("[{20 map[]}]"))
+				cache.Put(TestTSD{10, nil})
+				Expect(cache.String()).To(Equal("[{10 map[]} {20 map[]}]"))
+				cache.Put(TestTSD{40, nil})
+				Expect(cache.String()).To(Equal("[{10 map[]} {20 map[]} {40 map[]}]"))
+				cache.Put(TestTSD{50, nil})
+				Expect(cache.String()).To(Equal("[{20 map[]} {40 map[]} {50 map[]}]"))
+				cache.Put(TestTSD{30, nil})
+				Expect(cache.String()).To(Equal("[{30 map[]} {40 map[]} {50 map[]}]"))
+				cache.Put(TestTSD{50, nil})
+				Expect(cache.String()).To(Equal("[{40 map[]} {50 map[]} {50 map[]}]"))
+			})
+		})
+
+	})
+
+	Describe("Query", func() {
+		Context("when cache is empty", func() {
+			It("return empty results", func() {
+				result, ok := cache.Query(0, time.Now().UnixNano(), labels)
+				Expect(ok).To(BeTrue())
+				Expect(result).To(BeEmpty())
+			})
+		})
+		Context("when data put to cache do not execeeds the capacity", func() {
+			BeforeEach(func() {
+				capacity = 5
+			})
+			It("returns the data in [start, end)", func() {
+				cache.Put(TestTSD{20, nil})
+				result, ok := cache.Query(10, 40, labels)
+				Expect(ok).To(BeTrue())
+				Expect(result).To(Equal([]TSD{TestTSD{20, nil}}))
+
+				cache.Put(TestTSD{10, nil})
+				result, ok = cache.Query(10, 40, labels)
+				Expect(ok).To(BeTrue())
+				Expect(result).To(Equal([]TSD{TestTSD{10, nil}, TestTSD{20, nil}}))
+
+				cache.Put(TestTSD{40, nil})
+				result, ok = cache.Query(10, 40, labels)
+				Expect(ok).To(BeTrue())
+				Expect(result).To(Equal([]TSD{TestTSD{10, nil}, TestTSD{20, nil}}))
+
+				cache.Put(TestTSD{30, nil})
+				result, ok = cache.Query(10, 40, labels)
+				Expect(ok).To(BeTrue())
+				Expect(result).To(Equal([]TSD{TestTSD{10, nil}, TestTSD{20, nil}, TestTSD{30, nil}}))
+
+				cache.Put(TestTSD{50, nil})
+				result, ok = cache.Query(10, 40, labels)
+				Expect(ok).To(BeTrue())
+				Expect(result).To(Equal([]TSD{TestTSD{10, nil}, TestTSD{20, nil}, TestTSD{30, nil}}))
+			})
+		})
+
+		Context("when data put to cache execeed the capacity", func() {
+			BeforeEach(func() {
+				capacity = 3
+			})
+
+			Context("when all queried data are guarenteed  in cache", func() {
+				It("returns the data in [start, end)", func() {
+					cache.Put(TestTSD{20, nil})
+					cache.Put(TestTSD{10, nil})
+					cache.Put(TestTSD{40, nil})
+					cache.Put(TestTSD{30, nil})
+
+					result, ok := cache.Query(30, 50, labels)
+					Expect(ok).To(BeTrue())
+					Expect(result).To(Equal([]TSD{TestTSD{30, nil}, TestTSD{40, nil}}))
+
+					cache.Put(TestTSD{50, nil})
+					result, ok = cache.Query(35, 50, labels)
+					Expect(ok).To(BeTrue())
+					Expect(result).To(Equal([]TSD{TestTSD{40, nil}}))
```

Comment:
why query database directly, the "end" timestamp is included.  Is that possible to make the implementation compatible?

Reply:
Change made in https://github.com/cloudfoundry-incubator/app-autoscaler/pull/379/files#diff-689f77dbaf45e61a44746cb0b0ff27baR138 
mtrcs, ok := h.queryFunc(appId, start, end+1, order, labels)

use start, end+1 to query cache will make sure you get metrics data in [start, end]